genie-web/src/main/java/com/netflix/genie/web/tasks/job/JobCompletionService.java

```diff
@@ -0,0 +1,499 @@
+        this.archivalFailureRate = registry.counter("genie.jobs.archivalFailure.rate");
+        this.doneFileProcessingFailureRate = registry.counter("genie.jobs.doneFileProcessingFailure.rate");
+        this.finalStatusUpdateFailureRate = registry.counter("genie.jobs.finalStatusUpdateFailure.rate");
+        this.processGroupCleanupFailureRate = registry.counter("genie.jobs.processGroupCleanupFailure.rate");
+        this.archiveFileDeletionFailure = registry.counter("genie.jobs.archiveFileDeletionFailure.rate");
+        this.deleteDependenciesFailure = registry.counter("genie.jobs.deleteDependenciesFailure.rate");
+        // Retry template
+        this.retryTemplate = retryTemplate;
+    }
+
+    /**
+     * Event listener for when a job is completed. Updates the status of the job.
+     *
+     * @param event The Spring Boot application ready event to startup on
+     * @throws GenieException If there is any problem
+     */
+    void handleJobCompletion(final JobFinishedEvent event) throws GenieException {
+        final long start = System.nanoTime();
+        final String jobId = event.getId();
+        final Map<String, String> tags = Maps.newHashMap();
+
+        try {
+            final Job job = retryTemplate.execute(context -> getJob(jobId));
+
+            final JobStatus status = job.getStatus();
+
+            // Make sure the job isn't already done before doing something
+            if (status == JobStatus.INIT || status == JobStatus.RUNNING) {
+                try {
+                    retryTemplate.execute(context -> updateJob(job, event, tags));
+                } catch (Exception e) {
+                    log.error("Failed updating for job: {}", jobId, e);
+                    tags.put(ERROR_TAG, "JOB_UPDATE_FAILURE");
+                    finalStatusUpdateFailureRate.increment();
+                }
+                // Things that should be done either way
+                try {
+                    retryTemplate.execute(context -> processJobDir(job));
+                } catch (Exception e) {
+                    log.error("Failed archiving directory for job: {}", jobId, e);
+                    tags.put(ERROR_TAG, "JOB_DIRECTORY_FAILURE");
+                    archivalFailureRate.increment();
+                }
+                try {
+                    retryTemplate.execute(context -> sendEmail(jobId));
+                } catch (Exception e) {
+                    log.error("Failed sending email for job: {}", jobId, e);
+                    tags.put(ERROR_TAG, "SEND_EMAIL_FAILURE");
+                    emailFailureRate.increment();
+                }
+            }
+        } catch (Exception e) {
+            log.error("Failed getting job with id: {}", jobId, e);
+            tags.put(ERROR_TAG, "GET_JOB_FAILURE");
+        }
+        final Id timerId = this.jobCompletionId.withTags(tags);
+        this.registry.timer(timerId).record(System.nanoTime() - start, TimeUnit.NANOSECONDS);
```

Comment:
Why did you take this out of a finally block? If a runtime exception is thrown then it won't be executed